pulley/src/interp/interp_loop.rs

```diff
@@ -0,0 +1,656 @@
+}
+
+#[cfg(not(feature = "tail_calls"))]
+pub fn interpreter_loop(vm: &mut Vm, mut pc: UnsafeBytecodeStream) -> Done {
+    let opcode = Opcode::decode(&mut pc).unwrap();
+    let mut handler = OPCODE_HANDLER_TABLE[opcode as usize];
+
+    // As tight as we can get the interpreter loop without tail calls: while
+    // the handlers keep returning the next handler to call, call it.
+    loop {
+        match (handler.fun)(&mut vm.state, &mut pc) {
+            ControlFlow::Continue(next_handler) => handler = next_handler,
+            ControlFlow::Break(done) => return done,
+        }
+    }
+}
+
+#[cfg(feature = "tail_calls")]
+pub fn interpreter_loop(vm: &mut Vm, mut pc: UnsafeBytecodeStream) -> Done {
+    let opcode = Opcode::decode(&mut pc).unwrap();
+    let handler = OPCODE_HANDLER_TABLE[opcode as usize];
+
+    // The ideal interpreter loop: a bunch of opcode handlers tail calling
+    // each other!
+    match (handler.fun)(&mut vm.state, &mut pc) {
+        ControlFlow::Break(done) => done,
+    }
+}
+
+/// Wrap the business logic of each handler with the boilerplate of decoding
+/// operands and dispatching to next handler/exiting the loop.
+macro_rules! define_opcode_handlers {
+    (
+    $(
+        fn $name:ident (
+            $state:ident : &mut MachineState,
+            $pc:ident : &mut UnsafeBytecodeStream$(,)?
+            $($field:ident : $field_ty:ty),*
+        ) $body:block
+    )*
+    ) => {
+        $(
+            #[cfg(not(feature = "tail_calls"))]
+            pub fn $name($state: &mut MachineState, $pc: &mut UnsafeBytecodeStream) -> ControlFlow<Done, OpcodeHandler> {
+                let ($($field,)*) = crate::decode::operands::$name($pc);
+                match $body {
+                    ControlFlow::Continue(()) => {
+                        // Decode the next handler and return it so that `run`
+                        // can call it.
+                        let next_opcode = Opcode::decode($pc).unwrap();
+                        let next_handler = OPCODE_HANDLER_TABLE[next_opcode as usize];
+                        ControlFlow::Continue(next_handler)
+                    }
+                    ControlFlow::Break(done) => ControlFlow::Break(done),
+                }
+            }
+
+            #[cfg(feature = "tail_calls")]
+            pub fn $name($state: &mut MachineState, $pc: &mut UnsafeBytecodeStream) -> ControlFlow<Done, Uninhabited> {
+                let ($($field,)*) = crate::decode::operands::$name($pc);
+                match $body {
+                    ControlFlow::Continue(()) => {
+                        // Decode the next handler and return it so that `run`
+                        // can call it.
+                        let next_opcode = Opcode::decode($pc).unwrap();
+                        let next_handler = OPCODE_HANDLER_TABLE[next_opcode as usize];
+
+                        // FIXME: ICE
+                        // become (next_handler.fun)($state, $pc);
```

Comment:
And then we can actually use `become` here. I've tested locally and `cargo check` works with `become`, and avoids the ICE.

Reply:
Although maybe we want another `cfg(pulley_become)` or something so that if people want to rely on LLVM to optimize calls in tail position into tail calls, then they can?

Reply:
`cargo check` works because it doesn't actually generate LLVM/assembly, but if you try `cargo build` or `cargo test` it will

Reply:
Yep, that's the point: to be using `become` and still have `cargo check` type check the Rust code that we `cfg` on.

Not sure if it is actually *implemented* yet, but that should check that we aren't relying on any `Drop` implementations running after the tail call (which would preclude it from being an actual tail call instead of a call in tail position).

Reply:
So run `cargo check`/`cargo clippy` with `RUSTFLAGS="--cfg pulley_tail_calls"`, but not `cargo test`?

Reply:
Yeah just add a line something like

```yaml
    # Make sure that the tail-calls version of Pulley checks okay.
    - run: cargo check -p pulley-interpreter
      env:
        RUSTFLAGS: "--cfg pulley_tail_calls"
```

around here: https://github.com/bytecodealliance/wasmtime/blob/e38ffa197cf951ed513896fc63bb8e5c7b23135a/.github/workflows/main.yml#L499